### PR TITLE
New Days In Month Goodie (continues #2194)

### PR DESCRIPTION
**Pull request by 0xbsec**
Fixed and final version of pull request #2194
daysinmonth
IA page: https://duck.co/ia/view/days_in_month

### CODE REVIEW

**Comment by GuiltyDolphin** [comment]
@0xbsec Please [make an IA page](https://duck.co/ia/new_ia) and link to it in the PR description.

**Comment by 0xbsec** [comment]
@GuiltyDolphin Please check, Thanks.

File: lib/DDG/Goodie/DaysInMonth.pm [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+# Triggers - http://docs.duckduckhack.com/walkthroughs/calculation.html#triggers
+triggers any => 'how many days in','how many days are in', 'what is the number of days in', 'number of days in';
+
+# Handle statement
+handle remainder => sub {
+    my $remainder = $_;
+    return unless $remainder =~ qr/^\s*\w+\s*$/i;
+    my ($month) = $remainder =~ qr/(\w+)/;
+    return unless grep {$_ eq lc($month)} qw/january jan february feb march mar april apr may june jun july jul august aug september sep october oct november nov december dec/;
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DaysInMonth.pm]: Hmm... You should check out the `Dates` GoodieRole, it provides a bunch of regexes for months (as well as many other things).

Any date parsing should *really* be done by `Dates`.

Comment (0xbsec) [inline lib/DDG/Goodie/DaysInMonth.pm]: When Using `month_regex()` from `Dates` GoodieRole, the generated regex `(?^:(?^i:January|February|March|April|May|June|July|August|September|October|November|December)|(?^i:Jan|Feb|Mar|Apr|May|Jun|Jul|Aug|Sep|Oct|Nov|Dec))` will match `feb` as a month name from `febinvalid`, `invafeblid`, `invalidfeb`, which i don't want to happen in my case.

Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DaysInMonth.pm]: Just use boundaries.

```perl
my $month_triggers = month_regex();

my $month_re = qr/^$month_triggers$/;
```

Or similar.

Comment (0xbsec) [inline lib/DDG/Goodie/DaysInMonth.pm]: Please check now, thanks.

File: lib/DDG/Goodie/DaysInMonth.pm [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+# Handle statement
+handle remainder => sub {
+    my $remainder = $_;
+    return unless $remainder;
+    my $month_regex = month_regex();
+    my ($month) = $remainder =~ qr/^($month_regex)$/;
+    return unless $month;
+    my $days = calculateNumberOfDaysForMonthString($month);
+    return "Number of days in $month is $days.",
+    structured_answer => {
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DaysInMonth.pm]: Lets use a template, I think `text` is probably best.

File: lib/DDG/Goodie/DaysInMonth.pm [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        input => [ucfirst $month],
+        operation => 'Number of days in month',
+        result => ($days)
+    };
+};
+
+#Implemented the formula by Curtis Monroe
+#Original reference is found here: http://cmcenroe.me/2014/12/05/days-in-month-formula.html
+sub calculateNumberOfDays{
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DaysInMonth.pm]: What about leap years?

File: lib/DDG/Goodie/DaysInMonth.pm [hunk 4]
```diff
@@ -0,0 +1,65 @@
+package DDG::Goodie::DaysInMonth;
+# ABSTRACT: Returns number of days in a given month
+
+# Start at http://docs.duckduckhack.com/walkthroughs/calculation.html if you are new
```
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DaysInMonth.pm]: Lets remove all the introductory comments.

**Comment by moollaza** [comment]
@0xbsec Ping! I'm just checking in to see if you're still interested in finishing this one up? We'd love to see this go live!

There's absolutely no rush, just wanted to make sure you're not stuck on anything.

**Comment by daxtheduck** [comment]
Automated screenshot from [beta.duckduckgo.com](https://beta.duckduckgo.com/?q=how many days are in february)
                
                [Mobile screenshot](https://images.duckduckgo.com/iu/?u=https://ia-screenshots.s3.amazonaws.com/days_in_month_mobile.png)

File: lib/DDG/Goodie/DaysInMonth.pm (hunk 3 above)
Comment (GuiltyDolphin) [inline lib/DDG/Goodie/DaysInMonth.pm]: @0xbsec I'm not really happy with us doing calculation here - someone is bound to have done this before in a *robust* manner; we can just use a package.

I think the best thing to do would be to use DateTime and utilize the `last_day_of_month` method. Simple.

File: t/DaysInMonth.t [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+use warnings;
+use Test::More;
+use DDG::Test::Goodie;
+
+zci answer_type => "days_in_month";
+zci is_cached   => 1;
+
+ddg_goodie_test(
+    [qw( DDG::Goodie::DaysInMonth )],
+    'number of days in january' => test_zci(
```
Comment (GuiltyDolphin) [inline t/DaysInMonth.t]: We should clean up the tests - pull out the structured answer info into a sub and call that.

**Comment by GuiltyDolphin** [comment]
@0xbsec I'm closing this for now - if you want to finish it up later just let me know and we can re-open it.

**Pull request closed without merging**